### PR TITLE
fix typecheck in codegen

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Fix for #2706.

### How I did it

Extract type definition from node's `_metadata` if the type of the operands in a `BinOp` do not match in codegen.

### How to verify it

See test.

### Description for the changelog

Fix type-check for BinOp in codegen.

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/codegen/expr.py [hunk 1]
```diff
@@ -536,6 +536,14 @@ def parse_BinOp(self):
                     pos=pos,
                 )
 
+        # If the types do not match, check if type definition is propagated in node's metadata.
+        # This step is necessary for struct members.
+        if left.typ.typ != right.typ.typ:
+            if hasattr(self.expr.left, "_metadata") and "type" in self.expr.left._metadata:
```
Comment (charles-cooper) [inline vyper/codegen/expr.py:542]: there is no need to be so gentle here, we can go directly for `self.expr.left._metadata["type"]`. if the field is not there, it's a typechecker bug and we should fix it

File: vyper/codegen/expr.py [hunk 2]
```diff
@@ -536,6 +536,14 @@ def parse_BinOp(self):
                     pos=pos,
```
Comment (charles-cooper) [inline vyper/codegen/expr.py:536]: i think we should remove the above logic for handling literals and try to rely solely on the info passed down from the typechecker.

File: vyper/codegen/expr.py [hunk 3]
```diff
@@ -527,9 +532,16 @@ def parse_BinOp(self):
             return
 
         pos = getpos(self.expr)
-
-        left.typ = new_type_to_old_type(self.expr.left._metadata["type"])
-        right.typ = new_type_to_old_type(self.expr.right._metadata["type"])
+        left.typ = (
+            new_type_to_old_type(self.expr.left._metadata["type"])
+            if isinstance(self.expr.left, vy_ast.VyperNode)
```
Comment (charles-cooper) [inline vyper/codegen/expr.py:537]: why would `self.expr.left` ever not be an instance of `VyperNode`?

Comment (tserg) [reply vyper/codegen/expr.py:536]: this is a snippet from `examples/auctions/blind_auction.vy` where it is already converted to `IRNode` 

```
@external
@payable
def bid(_blindedBid: bytes32):
    # Check if bidding period is still open
    assert block.timestamp < self.biddingEnd

    # Check that payer hasn't already placed maximum number of bids
    numBids: int128 = self.bidCounts[msg.sender]
    assert numBids < MAX_BIDS

    # Add bid to mapping of all bids
    self.bids[msg.sender][numBids] = Bid({
        blindedBid: _blindedBid,
        deposit: msg.value
        })
    self.bidCounts[msg.sender] += 1
```

**Comment by tserg** [comment]
mostly works but have not refactored yet, and there is a namespace collision issue. 

currently, the annotation of the source code inserted by `sqrt` builtin is performed by injecting the variable names into the namespace. this causes namespace collision since the current names are commonly used (`x`, `y`, `z`). I think we could dynamically generate the variable names instead to ensure there are no clashes?

**Comment by charles-cooper** [comment]
> mostly works but have not refactored yet, and there is a namespace collision issue.
> 
> currently, the annotation of the source code inserted by `sqrt` builtin is performed by injecting the variable names into the namespace. this causes namespace collision since the current names are commonly used (`x`, `y`, `z`). I think we could dynamically generate the variable names instead to ensure there are no clashes?

probably should create a new, temporary namespace

File: vyper/codegen/expr.py (hunk 3 above)
Comment (charles-cooper) [reply vyper/codegen/expr.py:536]: i don't think that should be possible. all AST nodes are supposed to be VyperNodes; if we are seeing some that are not VyperNodes, it's indicative of a potential problem elsewhere.

File: vyper/codegen/expr.py [hunk 4]
```diff
@@ -203,17 +203,22 @@ def __init__(self, node, context):
 
     def parse_Int(self):
         # Literal (mostly likely) becomes int256
+        typ = (
+            new_type_to_old_type(self.expr._metadata["type"])
+            if "type" in self.expr._metadata
```
Comment (charles-cooper) [inline vyper/codegen/expr.py:208]: when do Int literals not have types attached to them?

File: vyper/codegen/stmt.py [hunk 5]
```diff
@@ -238,11 +236,10 @@ def _check_valid_range_constant(self, arg_ast_node, raise_exception=True):
         with self.context.range_scope():
             # TODO should catch if raise_exception == False?
             arg_expr = Expr.parse_value_expr(arg_ast_node, self.context)
-
         is_integer_literal = (
             isinstance(arg_expr.typ, BaseType)
-            and arg_expr.typ.is_literal
-            and arg_expr.typ.typ in {"uint256", "int256"}
+            and isinstance(arg_expr.value, int)
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:241]: you can use the `is_literal` property of IRnode -- `and arg_expr.is_literal`

File: vyper/codegen/stmt.py [hunk 6]
```diff
@@ -372,7 +368,12 @@ def _parse_For_list(self):
     def parse_AugAssign(self):
         target = self._get_target(self.stmt.target)
         sub = Expr.parse_value_expr(self.stmt.value, self.context)
-        if not isinstance(target.typ, BaseType):
+        target_typ = parse_type(
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:371]: when would `target_typ` not be equal to `target.typ`?

File: vyper/codegen/types/types.py [hunk 7]
```diff
@@ -397,12 +397,16 @@ def FAIL():
                 custom_structs,
             )
 
+        elif "type" in item._metadata:
```
Comment (charles-cooper) [inline vyper/codegen/types/types.py:400]: `parse_type` is supposed to be independent of the new type checking / annotation machinery. it's a good check that `parse_type` matches the type generated by the new type checker. if it doesn't match, it's indicative of a deeper problem and shouldn't be papered over like this.

**Comment by charles-cooper** [comment]
my thinking is that the codebase of the new type checker is not mature enough to support this yet, we need to refactor the new type checker and add more util functions so that there are not as many ad-hoc cases required for this PR.

**Review by fubuloubu — APPROVED** [review]

**Comment by charles-cooper** [comment]
@tserg i think we can close this now?

**Pull request closed without merging**